ibis/util.py

```diff
@@ -407,3 +407,56 @@ def wrapper(*args, **kwargs):
         return wrapper
 
     return decorator
+
+
+class Eq:
+    """A mixin for abstracting equality checks."""
+
+    __slots__ = ()
+
+    def __eq__(self, other: Any) -> bool:
+        return self.equals(other, cache={})
+
+    def __ne__(self, other: Any) -> bool:
+        return not (self == other)
+
+    def equals(
+        self,
+        other: Any,
+        cache: MutableMapping[Hashable, bool] | None = None,
+    ) -> bool:
+        return self is other or (
+            type(self) == type(other) and self._raw_equals(other, cache=cache)
+        )
+
+    def _raw_equals(
+        self,
+        other: Any,
+        cache: MutableMapping[Hashable, bool],
+    ) -> bool:
+        """Compare the individual fields of a subclass."""
+        return True
+
+
+class CachedEq(Eq):
```

Comment:
How about merging these classes? `cache` arg in `Eq._raw_equals`'s signature seems odd.

Reply:
They are separate to allow `Annotable` subclasses that aren't hashable. I'm not sure whether this occurs in practice but there are a few tests that use them.

Reply:
That shouldn't be possible asof the immutable enforcement PRs land.

Reply:
Makes sense, let's land that first then.

Reply:
Possibly requires a rebase but on top of that PR